Create data retention adr (#1459)
alphagov/email-alert-api: docs/adr/adr-007-data-retention.md

keep EmailAlertApi's database manageable going forward. There are a couple of
different approaches to consider, including storing unused data in cloud
storage away from the systems database (like we do when [archiving] emails) and
subsequently deleting the data locally. Or, setting up a means of removing the
data we don't use from the system entirely.

[archiving]: ...
[archived]: ...
[deleted]: ...
[nullified]: ...
[initial policy]: https://github.com/alphagov/email-alert-api/blob/master/docs/adr/adr-003-data-retention.md
[migrate subscribers to daily digest emails]: https://github.com/alphagov/email-alert-api/pull/1352

## Decision

We decided that historic, unused data should be removed periodically from the
system entirely. It was concluded that the data the system no longer uses
should only exist for a maximum of one year before removal. We sometimes
analyse historic data from up to a year ago but we have no evidence of needing
this data for analytics beyond one year. There is however some bespoke
behaviour for subscriber lists where we delete certain lists sooner than a year
but that will be elaborated on in a later section.

> We chose to remove the historic data entirely instead of retaining it through

Author: I've removed the paragraph in favour of the one you suggested in a later comment which sums up our approach to Athena.

archiving as the latter option seems unnecessary. If we've deemed the data as
historic and don't see a use for it (analytical or otherwise) going forward,
then retaining it via an external data store seems a waste of resources and
energy in maintaining the process.

It was also concluded that we don't need to change the retention for emails
(and subscription contents) as we already clean them up via the [email deletion

> mechanism] which is sufficient.

Author: This paragraph has been replaced by your suggestion later on so this comment is void.

The following list documents our approach to the decision for different data
in the system.

* **ContentChange, Message & DigestRun**. Remove once their `created_at` value is
  over a year old.

* **MatchedContentChange, MatchedMessage & DigestRunSubscriber**. Cascaded when
  their parent content change / message / digest run is deleted.

* **Subscription**. Remove when they've been ended for over a year, based on their
  `ended_at` value. Once a subscription is ended it [can't currently be made
  active again] and so can be deleted.

* **Previously subscribed to SubscriberList**. Remove lists when they have no
  active subscriptions and their latest, ended subscription is over a year old.
  It is possible for a list we deem as historic to be subscribed to again,
  however this seems unlikely due to the list not having any active
  subscriptions for over a year.  If this was to happen, after the list has
  been deleted then a new `SubscriberList` record will be generated.

> * **Unused SubscriberList**. Remove subscriber lists which have never had

Author: Hm we could, however, I think it reads nicer in a list as the list isn't about data with strictly one year retention.

> The following list documents our approach to the decision for different data
    in the system.

[deleted]: https://github.com/alphagov/email-alert-api/blob/master/app/workers/email_deletion_worker.rb
[initial policy]: https://github.com/alphagov/email-alert-api/blob/master/docs/adr/adr-003-data-retention.md
[migrate subscribers to daily digest emails]: https://github.com/alphagov/email-alert-api/pull/1352

## Decision

We decided that historic, unused data should be removed periodically from the
system entirely. It was concluded that the data the system no longer uses
should only exist for a maximum of one year before removal. We sometimes
analyse historic data from up to a year ago but we have no evidence of needing
this data for analytics beyond one year. There is however some bespoke
behaviour for subscriber lists where we delete certain lists sooner than a year
but that will be elaborated on in a later section.

We chose to remove the historic data entirely instead of retaining it through
archiving as the latter option seems unnecessary. If we've deemed the data as
historic and don't see a use for it (analytical or otherwise) going forward,
then retaining it via an external data store seems a waste of resources and
energy in maintaining the process.

> It was also concluded that we don't need to change the retention for emails

Reviewer: ```suggestion
Emails (and their associated SubscriptionContents) are the one place where we already manage retention - these are held locally for 7 days, and also archived to Athena indefinitely. We still need this low retention for Emails in order to manage the size of the table. However, the archiving to Athena is inconsistent with the above decision to delete unused data entirely. While we originally thought Athena might serve as a data source for analytics, this has not happened in practice, so the archive data stored there is entirely unused. We therefore decided to remove the archiving mechanism for Emails, since it had no value.
```

Reviewer: Saying this is sufficient contradicts the above decision and the context, where we've said that we won't hold unused data indefinitely. Although we could just be more specific and say "sufficient to keep the local database manageable", this is also the best time to clarify the future of archiving in Athena.

Sorry I didn't spot this earlier 😬. We'll need an extra consequence to reflect this. I'd suggest something like the following:

---

### Deprecate archiving Email to Athena

We planned work to remove it in the future. In the meantime, we consider archiving to Athena to be deprecated, and have updated [the documentation around it](https://github.com/alphagov/email-alert-api/blob/1a25d44126709bf8a98955d7609ced439095dd9c/docs/analytics.md) to reflect this status.

Author: > Saying this is sufficient contradicts the above decision and the context, where we've said that we won't hold unused data indefinitely.

This was referring to the email deletion mechanism, rather than the archiving aspect. Although they're interlinked so I can see this could be seen as contradictory.

I've updated to your recommendation and added the consequence.

(and subscription contents) as we already clean them up via the [email deletion
worker][deleted] which is sufficient.

The following list documents our approach to the decision for different data
in the system.

* **ContentChange, Message & DigestRun**. Remove once their `created_at` value is
  over a year old.

* **MatchedContentChange, MatchedMessage & DigestRunSubscriber**. Cascaded when
  their parent content change / message / digest run is deleted.

* **Subscription**. Remove subscriptions when they've been ended for over a
  year, based on their `ended_at` value. Once a subscription is ended it [can't
  currently be made active again] and so can be deleted.

* **Previously subscribed to SubscriberList**. Remove lists when they have no
  active subscriptions and their latest, ended subscription is over a year old.
  It is possible for a list we deem as historic to be subscribed to again,
  however this seems unlikely due to the list not having any active
  subscriptions for over a year.  If this was to happen, after the list has
  been deleted then a new `SubscriberList` record will be generated.

> * **Unused SubscriberList**. Remove lists which have never had any

Reviewer: Minor: we could make this a separate paragraph, since it's different to the other items in the list, which are all about retaining data for 1 year.

Author: Think this is a duplicate of a previous comment, I've responded to that one so resolving this one! 🤔

## Status

Accepted

## Consequences

### Historic deletion worker

We implemented a [worker] which runs periodically every day at [midday].
Removing this data was deemed as non urgent and so deleting once a day, in
hours, seemed sufficient.

### The first historic deletion run

On the first run of the worker which handles deleting historical data we
removed over 33 million records. A breakdown of the amount of deletions per
model can be found below:

* ContentChange - 148,617

* MatchedContentChange - 1,825,983

* Message - 15

* MatchedMessage - 16,099

* DigestRun - 677

* DigestRunSubscriber - 30,858,770

* Subscription - 1,046,576

* SubscriberList - 11,470

* Subscriber - 196,376

### Restrict analytics to 1 year of unused data

We now restrict our unused data for analytics to a period of at most 1 year.
This means any queries looking for this data (e.g. ended subscription's over a
year old) will be met with no results. We don't expect this to be an issue due
to our current expectation of the data we'll need for future analytics.

### Modify the nullify process and retire the deactivated concept

When documenting our data retention policy as described in this ADR, we found

> it difficult to understand and document the tangential concept of nullifying /

Reviewer: We could also link to [the original explanation](https://github.com/alphagov/email-alert-api/blob/master/docs/adr/adr-003-data-retention.md#subscriber-and-subscription) here.

deactivating subscribers.

To give some context on these concepts:

* Nullifying subscribers means to update the `address` of `Subscriber` records
  which have been unsubscribed for more than 28 days to `nil`.

* Deactivating subscribers is a concept that revolves around the
  `deactivated_at` field on the subscriber model. This field is populated when
  a subscriber unsubscribes from their last active subscription. It is used as
  a means to determine when a subscriber can be nullified.

We can however obtain the information held within `deactivated_at` through the

> `ended_at` field of the most recent ended subscription of the subscriber,

Reviewer: ```suggestion
`ended_at` field across all subscriptions of a subscriber,
```

We need to check that all subscriptions are ended, as well as determine the latest end time.

Author: I thought this was implied due to the nullify explanation referencing 'unsubscribed' subscribers. I've updated that explanation and this sentence to make it clearer.

meaning the deactivated concept is superfluous.

Subsequently we didn't want to spend energy documenting a process that didn't
make much sense to begin with so it was decided that it would be easier (and
beneficial as a whole) to just [remove] the deactivated concept completely.

In removing the deactivated concept we [modified the nullify process] to use

> the `ended_at` field of the last subscription, instead of `deactivated_at`. We

Reviewer: ```suggestion
the `ended_at` field, as described above, instead of `deactivated_at`. We
```
 
As above, we need to be more accurate here.

also updated the process to nullify subscribers which have no subscriptions
(active or ended) and are older than 28 days. This prevents the system from
storing user's personal data longer than needed, if for whatever reason, a

> subscriber ends up with no subscriptions. See examples of where this happened

Reviewer: ```suggestion
subscriber is created without any subscriptions. See examples of where this happened
```

We should be more precise to distinguish between a subscriber with "no (active) subscriptions" and a subscriber that has never had any subscriptions, which is what we mean here.

> # Decision Record: Data Retention

Reviewer 2: Looks like this is missing a number at the start, eg `7. Data Retention` as per [the template](https://github.com/alphagov/email-alert-api/blob/master/docs/adr/adr-005-record-architecture-decisions.md) and it's not necessary to state "Decision Record:".

I also notice [my ADR](https://github.com/alphagov/email-alert-api/pull/1476) is also flagged as number 7 🤦  I'll update mine to 8. Gosh it's been a while since I've been a project with concurrent ADR's - it's like being back on Publishing Workflow 😁 .

Author: Haha ADR's everywhere 👀 😆

> 

Reviewer 2: Looks like this is missing Date from the template: https://github.com/alphagov/email-alert-api/blame/master/docs/adr/adr-005-record-architecture-decisions.md#L3

## Context

> The original ADR concerning data retention resides [here][initial policy] and

Reviewer 2: I imagine it's worth adding a note in that ADR to say something like `This decision was revisited in [ADR-7](link) which supersedes this decision.`. I don't think you need to rename that file.

Author: Okay dokay, I've removed that commit and added a note in the previous ADR.

outlines our approach from February 2018. At that point in time it was decided

> we should retain all data produced by EmailAlertApi with the exception of

Reviewer: ```suggestion
we should retain all data in EmailAlertApi with the exception of
```

Minor: we still retained the data, but it was in a different place.

Reviewer 2: EmailAlertApi should be Email Alert API.

produce and store locally in EmailAlertApi. Consequently, this provoked worries
that perpetually keeping data could result in capacity issues due to the rapid
increase in data being stored. An example being we now generate ~430,000
`DigestRunSubscriber` records each day which is about eight times the amount
compared to that of ten months ago.

We've also gained more insight into what data we're not using for analytics.
The original plan documented in the previous retention policy was to retain as
much data as possible to enable us to do as much analysis as needed in the
future. However, much of this data is never used for analysis and so doesn't
serve a purpose in the email system.

Contrary to what was previously thought, storing lots of unused data
indefinitely actually complicates and pollutes our auditing and analysis. An
example of this can be seen if we look at subscriber lists. We have over
~73,000 lists with active subscribers but in the last week alone, we've
accumulated 1139 lists which don't have any subscriptions at all (ended or
active). These unused lists prevent us from doing simple auditing tasks
easily, like finding which lists to report on.

> Knowing the above, we thought about how we could cleanup this unused data to

Reviewer 2: ```suggestion
Knowing the above, we thought about how we could clean-up this unused data to
```

keep EmailAlertApi's database manageable going forward. There are a couple of
different approaches to consider, including storing unused data in cloud
storage away from the system's database (like we do when archiving emails) and
subsequently deleting the data locally. Or, setting up a means of removing the
data we don't use from the system entirely.

[archived]: https://github.com/alphagov/email-alert-api/blob/master/app/workers/email_archive_worker.rb
[deleted]: https://github.com/alphagov/email-alert-api/blob/master/app/workers/email_deletion_worker.rb
[initial policy]: https://github.com/alphagov/email-alert-api/blob/master/docs/adr/adr-003-data-retention.md
[migrate subscribers to daily digest emails]: https://github.com/alphagov/email-alert-api/pull/1352

## Decision

We decided that historic, unused data should be removed periodically from the
system entirely. It was concluded that the data the system no longer uses
should only exist for a maximum of 1 year before removal. We sometimes
analyse historic data from up to a year ago but we have no evidence of needing
this data for analytics beyond a year.

> Emails (and their associated SubscriptionContents) are the one place where we

Reviewer: Minor: this paragraph could go after the following list, which relates to the paragraph above.

already manage retention - these are held locally for 7 days, and also
archived to Athena indefinitely. We still need this low retention for Emails in
order to manage the size of the table. However, the archiving to Athena is
inconsistent with the above decision to delete unused data entirely. While we
originally thought Athena might serve as a data source for analytics, this has
not happened in practice, so the archive data stored there is entirely unused.
We therefore decided to remove the archiving mechanism for Emails, since it has
no value.

The following list documents our approach to the decision for different data
in the system.

* **ContentChange, Message & DigestRun**. Remove once their `created_at` value is
  over a year old.

> * **MatchedContentChange, MatchedMessage & DigestRunSubscriber**. Cascaded when

Reviewer 2: I'm not sure about the use of Cascaded as a verb - perhaps: Removed automatically

>   their parent content change / message / digest run is deleted.

Reviewer 2: ContentChange/Message/DigestRun would be consistent with the terminology you've used above.

* **Subscription**. Remove subscriptions when they've been ended for over a
  year, based on their `ended_at` value. Once a subscription is ended it [can't
  currently be made active again] and so can be deleted.

* **Previously subscribed to SubscriberList**. Remove lists when they have no

>   active subscriptions and their latest, ended subscription is over a year old.

Reviewer: ```suggestion
  subscriptions (after removing historic ones) and they are over a year old.
```

Minor: this is different to [the actual implementation](https://github.com/alphagov/email-alert-api/blob/master/app/workers/historical_data_deletion_worker.rb#L56).

Author: The implementation you've linked to is concerning historic subscribers, rather then historic subscriber lists so I've the wording as is. 👍

Reviewer: Oh sorry. I meant [here](https://github.com/alphagov/email-alert-api/blob/master/app/workers/historical_data_deletion_worker.rb#L46) - it is different.

Author: The ordering of the deletions was factored into the implementation to simply the code.

We decided that we should only remove subscriber lists which have no active subscriptions and the newest subscription is over a year old. As we delete the historic ended subscriptions prior to subscriber lists we don't have to include this extra logic in the code.

I could explain the way it was implemented but that seems to detract from the decision that was made. Does that make sense?

Reviewer: Yes, that's fair. It is just a minor point, and I wanted to make sure you were aware of the difference. I agree it's a valid perspective on the policy.

  It is possible for a list we deem as historic to be subscribed to again,
  however this seems unlikely due to the list not having any active
  subscriptions for over a year.  If this was to happen, after the list has
  been deleted then a new `SubscriberList` record will be generated.

* **Unused SubscriberList**. Remove lists which have never had any
  subscriptions and are over 7 days old. A scenario of when this can happen is
  when a user quits their signup journey midway through. 7 days was the chosen
  retention period to allow users to complete their signup journey as they have
  a [7 day limit] to confirm their subscription. As mentioned in the context
  section, unused subscriber lists pollute our analytics and so we should
  remove them regularly.

* **Subscriber**. Remove subscribers once their `created_at` is over a year old
  and they don't have any subscriptions (active or ended). Subscribers should
  only be without subscriptions when their historic subscriptions have been
  deleted. This means we should be able to delete all subscribers (under a year

>   old too) that don't have any subscriptions. However, we should be cautious

Reviewer 2: Minor - this area of caution seems somewhat superfluous to the decision. I'd suggest stopping the paragraph at "subscriptions."

Author: I've reworded the sentence so it relates more to the decision (as it was part of the decision). Let me know if thats any better?

  `deactivated_at` field on the subscriber model. This field is populated when
  a subscriber unsubscribes from their last active subscription. It is used as
  a means to determine when a subscriber can be nullified.

We can however obtain the information held within `deactivated_at` through the
`ended_at` field of the most recent ended subscription of an unsubscribed
user, meaning the deactivated concept is superfluous.

Subsequently we didn't want to spend energy documenting a process that didn't
make much sense to begin with so it was decided that it would be easier (and
beneficial as a whole) to just [remove] the deactivated concept completely.

In removing the deactivated concept we [modified the nullify process] to use
`ended_at`, as described above, instead of `deactivated_at`. We also updated
the process to nullify subscribers which have no subscriptions (active or
ended) and are older than 28 days. This prevents the system from storing user's
personal data longer than needed, if for whatever reason, a subscriber is
created without any subscriptions. See examples of where this happened
[here][sub bug] and [here][sub bug two].

### Deprecate archiving Email to Athena

> We planned work to remove it in the future. In the meantime, we consider

Reviewer: ```suggestion
We have planned work to remove it in the future. In the meantime, we consider
```

# Decision Record: Data Retention

## Context

The original ADR concerning data retention resides [here][initial policy] and
outlines our approach from February 2018. At that point in time it was decided
we should retain all data produced by EmailAlertApi with the exception of
emails which are [archived] to S3 & [deleted] after a week.

Since then we've come to realise that a revision of our data retention policy

> is needed as much of the data we retaine will never be used by us, or the

Author: Sneaky 'e' suffix trying to make an appearance 😆 Thanks for spotting!

example of this can be seen if we look at subscriber lists. We have over
~73,000 lists with active subscribers but in the last week alone, we've
accumulated 1139 lists which don't have any subscriptions at all (ended or
active). These unused lists prevent us from doing doing simple auditing tasks
easily, like determining which lists are actually subscribed to.

Knowing the above, we thought about how we could cleanup this unused data to
keep EmailAlertApi's database manageable going forward. There are a couple of
different approaches to consider, including storing unused data in cloud
storage away from the systems database (like we do when [archiving] emails) and
subsequently deleting the data locally. Or, setting up a means of removing the
data we don't use from the system entirely.

[archiving]: ...
[archived]: ...
[deleted]: ...
[nullified]: ...
[initial policy]: https://github.com/alphagov/email-alert-api/blob/master/docs/adr/adr-003-data-retention.md
[migrates subscribers to daily digest emails]: https://github.com/alphagov/email-alert-api/pull/1352

## Decision

We decided that historic (unused) data should be removed periodically from the

> system entirely, instead of retaining it through archiving. Archiving was

Author: Hmm I didn't say it did mean useless. I also said that the data wasn't used after a point in the next paragraph.

However, after your comment I've reshuffled both paragraphs which should make it clearer that we don't need historic data after a certain point.

deemed as unnecessary because if we've deemed data as being historic then we
don't see a use for it (analytical or otherwise) going forward and so retaining
it via an external data store seems a waste of resources and energy in
maintaining the process.

> As part of the decision we concluded that the data we, and the system no longer

Author: I mentioned that we don't have purpose for this data beyond a year and thought that implicitly implied that we do have a purpose for data under a year old. 

I've been more explicit anyways which should make it clearer. 👍

uses should only exist for a maximum of one year before removal due to there
being no evidence of needing unused data for analytics beyond one year. There
is however some bespoke behaviour for subscriber lists where we delete certain
lists sooner than a year but that will be elaborated on in a later section.

We also concluded that we don't need to change the retention for emails (and
subscription contents) as we already clean them up via the [email deletion
mechanism] which is sufficient.

The following sections document our approach to the decision for different data
in the system.

> ### Data which has previously been used

Reviewer: I'm finding the formatting in this section quite hard to follow, especially with the hrule, which cuts the whole document in half.

I appreciate there are quite a few levels at this point. Something which might work better would be to use a list, like this:

- **ContentChange, Message & DigestRun**. Remove once their created_at value is over a year old.

...

- **SubscriberList**. We remove lists when they have no active subscriptions and their latest, ended subscription is over a year old. It is possible for a list we deem as historic to be subscribed to again, however this seems unlikely due to the list not having any active subscriptions for over a year. If this was to happen after the list has been deleted then a new SubscriberList record will be generated.

Author: Arghglhjsncdp I did this originally. 😆 

I've used a list which does make the section easier to read.

Let me know if it's easier to follow! 👍

includes:

##### Subscription

We remove subscriptions when they've been ended for over a year, based on their
`ended_at` value. Once a subscription is ended it [can't currently be made
active again] and so can be deleted.

##### SubscriberList

We remove lists when they have no active subscriptions and their latest, ended
subscription is over a year old.

It is possible for a list we deem as historic to be subscribed to again,
however this seems unlikely due to the list not having any active subscriptions
for over a year. If this was to happen after the list has been deleted then a
new `SubscriberList` record will be generated.

##### Subscriber

We remove subscribers once their `created_at` is over a year old and they don't
have any subscriptions (active or ended).

Subscribers should only be without subscriptions (active or ended) when their
historic subscriptions have been deleted via the historic deletion worker. This
means we should be able to delete all subscribers (under a year old too) that
don't have any subscriptions (active or ended).

However, we should be cautious about doing this so we don't accidently remove

> new subscribers when the application changes. Some examples of where it was

Reviewer: ```suggestion
new subscribers that lack subscriptions, due to flaws in our code. Some examples of where it was
```

We should be more specific about what we're worried about here.

unintentionally possible to end up with subscribers with no subscriptions can
be found [here][sub bug] and [here][sub bug two].

### Data which has never been used

We remove subscriber lists which have never had any subscriptions and are over
seven days old. A scenario of when this can happen is when a user quits their
signup journey midway through. Seven days was the chosen retention period to
allow users to complete their signup journey as they have a [seven day limit]
to confirm their subscription.

> As mentioned in the context section, subscriber lists pollute our analytics and

Reviewer: ```suggestion
As mentioned in the context section, unused subscriber lists pollute our analytics and
```

so we should remove them regularly.

[can't currently be made active again]: https://github.com/alphagov/email-alert-api/blob/master/docs/adr/adr-003-data-retention.md#subscriber-and-subscription
[seven day limit]: https://github.com/alphagov/email-alert-api/blob/f0c61539f7244bdb52db54d277a46412b728d642/app/services/auth_token_generator_service.rb#L7

## Status

Accepted

## Consequences

### Historic deletion worker

We implemented a [worker] which runs periodically every day at [midday].
Removing this data was deemed as non urgent and so deleting once a day, in
hours, seemed sufficient.

### The first historic deletion run

On the first run of the worker which handles deleting historical data we
removed over 33 million records. A breakdown of the amount of deletions per
model can be found below:

* ContentChange - 148,617

* MatchedContentChange - 1,825,983

* Message - 15

* MatchedMessage - 16,099

* DigestRun - 677

* DigestRunSubscriber - 30,858,770

* Subscription - 1,046,576

* SubscriberList - 11,470

* Subscriber - 196,376

> ### Presence of emails during historic deletion

Reviewer: The presence of emails isn't something new. I think this is really an implementation detail, and a pretty minor one at that. We could combine it with the section about the worker, and maybe comment on how this may have once been a problem, but [ADR 006](https://github.com/alphagov/email-alert-api/blob/master/docs/adr/adr-006-email-delivery-responsibilities.md) means it's basically impossible now.

Author: > I think this is really an implementation detail, and a pretty minor one at that

Hmm true, I've chose not to include it at all given it's deemed as 'minor'.

If an old email exists and belongs to a historic subscriber then the historic
worker will throw an error when the subscriber deletion is attempted. This is
due to a db restriction as these emails should've been cleaned up already via
the [email deletion mechanism]. We want to produce an error for this scenario
rather than just delete them so bugs don't unintentionally creep in. This also
applies to subscription contents as their deletion is cascaded when deleting
emails.

> ### Restricts analytics to 1 year of unused data

Reviewer: ```suggestion
### Restrict analytics to 1 year of unused data
```

> We restrict our unused data for analytics to a period of at most 1 year. This

Reviewer: ```suggestion
We now restrict our unused data for analytics to a period of at most 1 year. This
```

Minor, but this should help the reader understand what's changing.

means any queries looking for this data (e.g. ended subscriptions over a year
old) will be met with no results. We don't expect this to be an issue but is

> something worth mentioning.

Reviewer: Why is it worth mentioning? Perhaps we could emphasise why we don't expect it to be an issue.

### Modify the nullify process and retire the deactivated concept

> When documenting our data retention policies we re-discovered a concept which

Reviewer: We should clarify why this section is in the ADR. It's true that we re-discovered the concept, but that alone doesn't merit a section here. As I recall, the problem was that we found it hard to write documentation for the retention policy as described in this ADR, due to a tangential concept of nullifying / deactivating subscribers, which turned out to be hard to understand, and thus, to document.

Author: Yep true we did figure out it was hard to document. I've updated the intro to include this.

didn't make much sense regarding the activation / deactivation of subscribers.
This concept was a point of confusion in the email system and wasn't actually
needed so has been [removed].

The concept revolved around the `deactivated_at` field on the subscriber model. This
field was used as a means of determining when a subscriber could be nullified.
We can however obtain this information via the most recent ended subscription

> of the subscriber (based on the subscriptions `ended_at`) which means

Reviewer: ```suggestion
of the subscriber (based on the subscription's `ended_at`) which means
```
